### DISCUSSION TITLE
Fixing a Bug in a Simple Calculator

### DISCUSSION THREAD

**Discussion by eFutureDev** [community/community#110203] [Programming Help] [answered]
### Select Topic Area

Bug

### Feature Area

Issues

### Body

I'm building a basic calculator using HTML, CSS, and JavaScript. Users can input two numbers and select an operation (addition, subtraction, multiplication, or division), and the calculator should display the result when the "Calculate" button is clicked. However, users are reporting that the calculator is not giving correct results for division operations. Find and fix the bug in the division operation.

```
<label for="num1">Number 1:</label>
<input type="number" id="num1"><br>

<label for="num2">Number 2:</label>
<input type="number" id="num2"><br>

<label for="operation">Operation:</label>
<select id="operation">
  <option value="add">Addition</option>
  <option value="subtract">Subtraction</option>
  <option value="multiply">Multiplication</option>
  <option value="divide">Division</option>
</select><br>

<button onclick="calculate()">Calculate</button><br>

<label for="result">Result:</label>
<input type="text" id="result" disabled>

function calculate() {
    var num1 = parseFloat(document.getElementById("num1").value);
    var num2 = parseFloat(document.getElementById("num2").value);
    var operation = document.getElementById("operation").value;
    var resultField = document.getElementById("result");

    var result;
    if (operation === "add") {
        result = num1 + num2;
    } else if (operation === "subtract") {
        result = num1 - num2;
    } else if (operation === "multiply") {
        result = num1 * num2;
    } else if (operation === "divide") {
        result = num1 / num2; // There's a bug here!
    }

    resultField.value = result;
}
```

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to the GitHub feature areas [Issues](https://docs.github.com/en/issues/tracking-your-work-with-issues/about-issues) and/or [Projects](https://docs.github.com/en/issues/planning-and-tracking-with-projects/learning-about-projects/about-projects).

**Comment by Lanrad23** [comment, marked as answer]
The issue with the division part of your calculator might not be obvious from what you've shown. The way you've set it up to divide numbers (result = num1 / num2;) looks right. But, problems can happen if someone tries to divide by zero or because JavaScript sometimes gets a bit tricky with numbers.
Also, make sure the option for division in your calculator matches exactly in both your website's form and the JavaScript code. For example, if the choice for dividing is written as <option value=""divide"">Division</option> in the form, your JavaScript needs to use ""divide"" to know when to do the division. If these don't match up perfectly, the division won't work right.

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @eFutureDev ! 

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Projects and Issues category is solely related to conversations around the GitHub products Issues and Projects. 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information. Note:  we ask that members contribute in a positive and constructive way. Posting homework or test questions without context will likely discourage engagement and responses.

If this content is posted again, your Community account will be subject to immediate and indefinite suspension.
For additional guidance, please review our [Community Code of Conduct](https://github.com/community/community/blob/main/CODE_OF_CONDUCT.md) and [GitHub’s Terms of Service](https://docs.github.com/en/site-policy/github-terms/github-terms-of-service).

Thank you!

**Comment by ethanappgenius** [comment]
The issue with the division operation in your calculator lies in the fact that division by zero is not handled. When users input zero as the second number, it results in a division by zero error, causing incorrect results or unexpected behavior. To fix this bug, you should add a check to ensure that the second number is not zero before performing the division operation. You can do this by adding a conditional statement to verify if num2 is not equal to zero before executing the division operation. If num2 is zero, you can display an error message or handle the situation according to your requirements to prevent the calculation from proceeding. This adjustment will ensure that the calculator provides accurate results for all division operations.

**Comment by larahkaia53-commits** [comment]
Nice explanation of fixing a calculator bug! For anyone looking for ready-to-use online calculators, check out [CalculatorUp.com](https://calculatorup.com/)
 — it’s free and covers math, finance, health, and more.

**Comment by tipcalculator0** [comment]
Looks like the main bug is in the division block — when num2 is 0, the calculator still tries to divide, which leads to incorrect results (Infinity or NaN). You should add a check before performing division, for example:

} else if (operation === "divide") {
    if (num2 === 0) {
        result = "Error: Cannot divide by zero";
    } else {
        result = num1 / num2;
    }
}


That way your calculator will return a meaningful message instead of breaking. I faced a similar issue when testing some logic on https://tipcalculator.us/
, and adding input validation fixed it.
